### PR TITLE
[Bugfix] Zero stale is_prefilling in padded CUDA graph rows for Mamba

### LINKED ISSUE #41841: Stale padded request metadata can misclassify Mamba CUDA graph rows
### Summary

Full CUDA graph metadata can include stale CPU request-state values for padded rows. For Mamba backends this can make padding rows look like prefills, so a decode-only CUDA graph replay can be classified as mixed decode/prefill metadata.

The relevant path is:

- `GPUModelRunner._build_attention_metadata` slices `num_computed_tokens_cpu_tensor[:num_reqs_padded]` and `num_prompt_tokens_cpu_tensor[:num_reqs_padded]`.
- Other padded metadata is explicitly neutralized, e.g. padded block table entries are filled with `NULL_BLOCK_ID` and padded sequence lengths are zeroed.
- The request-state tensors are not neutralized beyond `num_reqs`.
- `is_prefilling = num_computed_tokens_cpu < num_prompt_tokens_cpu` can therefore be true for inactive padded rows if those rows contain stale values from prior requests.
- Mamba calls `split_decodes_and_prefills(..., treat_short_extends_as_decodes=False)`, which ORs `is_prefilling` into the split. A zero-token padded row can become a fake prefill.

### Minimal Repro

This repro is synthetic and does not require a model or dataset. It demonstrates the metadata classification issue directly.

```python
import torch

from vllm.v1.attention.backend import CommonAttentionMetadata
from vllm.v1.attention.backends.utils import split_decodes_and_prefills

num_reqs = 2
num_reqs_padded = 4

# Two real decode rows, then two padded rows with zero query length.
query_start_loc = torch.tensor([0, 1, 2, 2, 2], dtype=torch.int32)
seq_lens = torch.tensor([8, 9, 0, 0], dtype=torch.int32)

# Real rows are decode-only, but padded rows contain stale state from older requests.
num_computed_tokens = torch.tensor([7, 8, 1, 2], dtype=torch.int32)
num_prompt_tokens = torch.tensor([4, 4, 3, 4], dtype=torch.int32)

is_prefilling = num_computed_tokens[:num_reqs_padded] < num_prompt_tokens[:num_reqs_padded]
metadata = CommonAttentionMetadata(
    query_start_loc=query_start_loc,
    query_start_loc_cpu=query_start_loc,
    seq_lens=seq_lens,
    num_reqs=num_reqs_padded,
    num_actual_tokens=2,
    max_query_len=1,
    max_seq_len=9,
    block_table_tensor=torch.zeros((num_reqs_padded, 1), dtype=torch.int32),
    slot_mapping=torch.arange(2, dtype=torch.int64),
    is_prefilling=is_prefilling,
)

print(split_decodes_and_prefills(metadata, treat_short_extends_as_decodes=False))
```

Current output:

```text
(2, 2, 2, 0)
```

Expected output for a padded decode-only full CUDA graph batch:

```text
(4, 0, 2, 0)
```

If the padded CPU request-state rows are cleared before computing `is_prefilling`, the split returns the expected decode-only result:

```python
fixed_num_computed = num_computed_tokens[:num_reqs_padded].clone()
fixed_num_prompt = num_prompt_tokens[:num_reqs_padded].clone()
fixed_num_computed[num_reqs:num_reqs_padded] = 0
fixed_num_prompt[num_reqs:num_reqs_padded] = 0
fixed_metadata = metadata.replace(
    is_prefilling=fixed_num_computed < fixed_num_prompt,
)
print(split_decodes_and_prefills(fixed_metadata, treat_short_extends_as_decodes=False))
# (4, 0, 2, 0)
```

### Why this matters

This only affects padded CUDA graph metadata rows. It is most likely to surface in workloads that combine:

- Mamba or hybrid Mamba models
- full CUDA graph replay with padded request rows
- prefix caching / Mamba align mode or any path sensitive to Mamba decode-vs-prefill metadata
- long decode batches where request rows churn and stale inactive-row state is more likely to matter

### Possible Fix

When building common attention metadata with `num_reqs_padded > num_reqs`, clone and zero `num_computed_tokens_cpu` and `num_prompt_tokens_cpu` for padded rows before computing `is_prefilling`.

### PR DESCRIPTION
**Pull request by liulanze**
## Purpose

Fix stale `is_prefilling` metadata in padded CUDA graph rows that misclassifies Mamba decode requests as prefills.

After `condense()` compacts active requests, vacated slots retain stale `num_computed_tokens` and `num_prompt_tokens` values. When the batch is padded for CUDA graphs, the comparison `num_computed < num_prompt` can yield `True` for padding rows. Mamba is the only backend that passes `treat_short_extends_as_decodes=False` to `split_decodes_and_prefills`, so these stale `True` values get OR-ed into `is_prefill` and shift the decode/prefill boundary, misclassifying real decode rows as prefills.

Fix: zero `is_prefilling[num_reqs:]` immediately after computation.

Resolves #41841

## Test Plan

```bash
pytest tests/v1/attention/test_attention_splitting.py -xvs
```

## Test Result

```
23 passed, 20 warnings in 5.49s
```

- `test_split_decodes_stale_padded_is_prefilling` — verifies correct splitting after zeroing padded rows
- `test_split_decodes_stale_padded_is_prefilling_without_fix` — confirms stale data shifts the boundary without the fix

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by liulanze** [comment]
Hi @njhill, just wanted to gently ping this PR. The fix is minimal and the PR includes tests that demonstrate both the bug and the fix. Happy to address any feedback or questions. Thanks!

**Comment by njhill** [comment]
Thanks @liulanze, LGTM! Would be good to also get nod from @tdoublep.

**Comment by liulanze** [comment]
The `metrics-tracing-2-gpus` failure is a segfault in a background thread during OpenTelemetry/gRPC shutdown after the test completed successfully, it's unrelated to this PR's changes. Maybe could we re-run this job?

File: tests/v1/attention/test_attention_splitting.py [hunk 1]
```diff
@@ -379,3 +379,69 @@ def test_prefill_split_across_ubatches(
...
+    common.is_prefilling = is_prefilling
+
+    num_decodes, num_prefills, num_decode_tokens, num_prefill_tokens = (
+        split_decodes_and_prefills(
+            common,
+            decode_threshold=1,
+            require_uniform=False,
+            treat_short_extends_as_decodes=False,
+        )
+    )
```
Comment (tdoublep) [inline tests/v1/attention/test_attention_splitting.py]: imo this test doesn't really "test the fix" but it just tests that `split_decodes_and_prefills` does what it should do. Do we have existing tests for the latter? I think a real test should somehow involve checking that `_build_attention_metadata` does the correct thing when stale data exists.

Comment (liulanze) [reply tests/v1/attention/test_attention_splitting.py]: Updated the test to call `_build_attention_metadata` directly with stale data in the padded rows, and assert that those rows are zeroed in the resulting `CommonAttentionMetadata`.

Validated on an A100: test fails on stock vLLM (rows 3–4 remain `True`) and passes after the fix.

Please review again, thanks.

**Review by tdoublep — APPROVED** [review]
LGTM
